[16.0][REF] l10n_br_fiscal, l10n_br_account: Cálculo e Visualização das Alíquotas Efetivas do Simples Nacional

antoniospneto opened the pull request:
Migração da PR https://github.com/OCA/l10n-brazil/pull/2822 (14.0)

antoniospneto committed:
[REF] l10n_br_account: simples nacional taxes

antoniospneto committed:
[FIX] l10n_br_nfe: simples nacional xml test

rvalyi commented:
@antoniospneto vc viu o erro na NFe do Simples né?

```
 2025-03-25 02:48:51,760 437 INFO odoo odoo.addons.l10n_br_nfe.tests.test_nfe_serialize: XML file saved at /github/home/.local/share/Odoo/filestore/odoo/34/343543298062bbed0aaffd7f601a7358f967a3d1 
2025-03-25 02:48:51,776 437 INFO odoo odoo.addons.l10n_br_nfe.tests.test_nfe_serialize_sn: Diff with expected XML (if any): [UpdateTextIn(node='/*/*/*[4]/*[2]/*[1]/*/*[4]', text='0.37')] 
2025-03-25 02:48:51,776 437 INFO odoo odoo.addons.l10n_br_nfe.tests.test_nfe_serialize_sn: ====================================================================== 
2025-03-25 02:48:51,776 437 ERROR odoo odoo.addons.l10n_br_nfe.tests.test_nfe_serialize_sn: FAIL: TestNFeExportSN.test_serialize_xml
Traceback (most recent call last):
  File "/__w/l10n-brazil/l10n-brazil/l10n_br_nfe/tests/test_nfe_serialize_sn.py", line 24, in test_serialize_xml
    assert len(diff) == 0
AssertionError
```

Na v16 o Odoo remove os attachments criados no testes. Ou vc comenta isso no framework do teste do Odoo, ou vc da um print do XML e abre no meld para investigar e resolver.

cc @marcelsavegnago teria que usar a mesma tecnica para finalizar a migração do l10n_br_cte até na v16: https://github.com/OCA/l10n-brazil/pull/3663#issuecomment-2744696970

antoniospneto commented:
@rvalyi valeu, vou dar uma olhada nisso, eu tenho um caso de teste que preciso migrar do módulo fiscal pro fiscal_edi tbm, depois revisar umas coisas que o Renato tinha pedido na 14.0.

rvalyi commented:
Aqui a gente ta ficando menos afogado nas migrações para a v16, então o @renatonlima deve conseguir participar mais agora tb.

rvalyi commented:
@antoniospneto vejo que os testes estão verdes, mas ficou pronto para revisar? Pode dar um rebase tb por favor?

antoniospneto commented:
@rvalyi ainda vou dar mais uma revisada no código antes

Alguem entende pq o pre-commit fica toda hora removendo e adicionando esses espaços no README?

![image](https://github.com/user-attachments/assets/10fb8061-3435-4c12-ab64-6bf473980f49)

O Diff sempre cresce por causa disso :/

rvalyi commented:
@antoniospneto eu não sei mas eu nunca jogo o README.rst no commit para não atrapalhar com isso... (a não ser que for um modulo novo).

rvalyi commented:
@antoniospneto talvez vc não tem a mesma versão de alguma biblioteca que converte do markdown para rst (docutils?) do que no bot da OCA e ai fica essas discrepâncias...

antoniospneto commented:
@rvalyi deu certo agora, apaguei os arquivos do readme gerados automaticamente pelo pre-commit, fiz o commit novamente com a opção "-n"

antoniospneto commented:
Pessoal, pronto pra revisão :)

marcelsavegnago commented:
ping @WesleyOliveira98 sugiro fazer teste funcional até para servir de aprendizado.

antoniospneto commented:
@renatonlima o que você acha, tranquilo seguir com essa pr?

renatonlima commented:
Olá @antoniospneto valeu pela contribuição:

Alguns pontos sobre o PR:

Eu vi que você adicionou um objeto l10n_br_fiscal.simplified.tax.effective para ter o calculo da alíquota efetiva por empresa com base nas faixas de cada anexo. Na minha opinião eu não acho necessário ter mais um objeto para fazer esse calculo e ter mais de um anexo do Simples Nacional (Em alguns casos a empresa tem mais de um anexo do SN)

Minha sugestão seria ter um campo m2m no res.company com o objeto l10n_br_fiscal.simplified.tax e ser calculado levando em consideração os campos to regime fiscal (simples nacional), CNAEs primário, secundário e o campo is_industry da empresa para determinar em quais anexos do simples a empresa se enquadra porque uma empresa pode se enquadrar em mais de 1 anexo.

Existe um problema também que a RBT12 (receita bruta nos últimos 12 meses) que quando uma empresa tem mais de um anexo a apuração da RBT12 deve ser apurada separada por cada anexo (por isso eu tinha colocado o tax_range_id) na linha dos documentos fiscais para saber depois o faturamento de cada anexo separado, apesar de não estar implementado essa funcionalidade, mas seria bom manter esse campo para ter essa informação quando a apuração for implementada.

Eu vou fazer outro comentário detalhando melhor uma sugestão para implementação.

marcelsavegnago commented:
cc @WesleyOliveira98

antoniospneto commented:
Olá @renatonlima 

> Eu vi que você adicionou um objeto l10n_br_fiscal.simplified.tax.effective para ter o calculo da alíquota efetiva por empresa com base nas faixas de cada anexo. Na minha opinião eu não acho necessário ter mais um objeto para fazer esse calculo e ter mais de um anexo do Simples Nacional (Em alguns casos a empresa tem mais de um anexo do SN)

Essa é apenas uma abordagem técnica alternativa para estruturar e calcular os dados. A vantagem do modelo que propus é que ele segue uma separação clara de responsabilidades entre os objetos. Ainda assim, a solução que implementei suporta perfeitamente empresas com múltiplos anexos do Simples Nacional.

> Minha sugestão seria ter um campo m2m no res.company com o objeto l10n_br_fiscal.simplified.tax e ser calculado levando em consideração os campos to regime fiscal (simples nacional), CNAEs primário, secundário e o campo is_industry da empresa para determinar em quais anexos do simples a empresa se enquadra porque uma empresa pode se enquadrar em mais de 1 anexo.

No meu caso, o cálculo considera todos os anexos possíveis, mesmo que a empresa não esteja enquadrada em todos. Isso porque, na linha da fatura, o usuário não define o CNAE do produto/serviço — então uso o CFOP como critério para identificar o anexo aplicável. Vale observar que estou focando apenas na lógica de produtos (anexo I e II); ainda não estou tratando os anexos relacionados a serviços.

> Existe um problema também que a RBT12 (receita bruta nos últimos 12 meses) que quando uma empresa tem mais de um anexo a apuração da RBT12 deve ser apurada separada por cada anexo (por isso eu tinha colocado o tax_range_id) na linha dos documentos fiscais para saber depois o faturamento de cada anexo separado, apesar de não estar implementado essa funcionalidade, mas seria bom manter esse campo para ter essa informação quando a apuração for implementada.

Não entendi muito bem essa parte. A legislação do Simples estabelece que a alíquota deve ser aplicada considerando a receita bruta total da empresa, e não segregada por anexo. Você poderia confirmar se essa separação que mencionou é realmente obrigatória na apuração fiscal? Ou seria apenas para fins gerenciais?

mileo commented:
[Apuração de Receita e Alíquotas Efetivas no Simples Nacional com Múltiplos Anexos.pdf](https://github.com/user-attachments/files/20212659/Apuracao.de.Receita.e.Aliquotas.Efetivas.no.Simples.Nacional.com.Multiplos.Anexos.pdf)

antoniospneto committed:
[REF] l10n_br_fiscal: simples nacional taxes

antoniospneto commented:
feito rebase, foi removido uns onchanges dos testes, pois com as aultimas atualizações não há mais necessidade.

rvalyi commented:
pessoal, o @renatonlima deve subir umas coisas a respeito disso, ai teria que ver porque talvez iria invalidar essa proposta.
